### PR TITLE
fix ConcurrentModificationException while PlatformMetricsRegistry stop

### PR DESCRIPTION
**Pull request by samarth-gupta-traceable**
fix ConcurrentModificationException while PlatformMetricsRegistry stop

### CODE REVIEW

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java [hunk 1]
```diff
@@ -37,6 +37,7 @@
 import java.time.Duration;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.Iterator;
```
Comment (tim-mwangi) [inline platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:40]: who is using this import?

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java [hunk 2]
```diff
@@ -316,10 +317,10 @@ public static synchronized void stop() {
...
     DEFAULT_TAGS.clear();
     METER_REGISTRY.forEachMeter(METER_REGISTRY::remove);
-    METER_REGISTRY.getRegistries().forEach(e -> {
-      e.clear();
-      METER_REGISTRY.remove(e);
-    });
+    METER_REGISTRY.getRegistries().forEach(MeterRegistry::clear);
+    Set<MeterRegistry> registries = new HashSet<>(METER_REGISTRY.getRegistries());
+    registries.forEach(METER_REGISTRY::remove);
+    registries.clear();
```
Comment (avinashkolluru) [inline platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:323]: I'm confused. The clear is on the new HashSet object you are creating. So the underlying `METER_REGISTRY.getRegistries()` still has the registries right?

Comment (avinashkolluru) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:323]: Though we got rid of the problem, if the intention is to clear those registries I think that wouldn't happen

Comment (samarth-gupta-traceable) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:323]: @avinashkolluru  `registries.forEach(METER_REGISTRY::remove)`  will remove it from METER_REGISTRY.remove()
Screenshot 2020-08-26 at 11 31 46 PM

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java (hunk 1 above)
Comment (samarth-gupta-traceable) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java]: removed

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java (hunk 2 above)
Comment (avinashkolluru) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:322]: Dumb. Sorry.

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java [hunk 3]
```diff
@@ -316,10 +316,10 @@ public static synchronized void stop() {
 
     DEFAULT_TAGS.clear();
     METER_REGISTRY.forEachMeter(METER_REGISTRY::remove);
-    METER_REGISTRY.getRegistries().forEach(e -> {
-      e.clear();
-      METER_REGISTRY.remove(e);
-    });
+    METER_REGISTRY.getRegistries().forEach(MeterRegistry::clear);
```
Comment (avinashkolluru) [inline platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:319]: As discussed offline, Should we call a `close` too before clearing?

Comment (samarth-gupta-traceable) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:319]: @buchi-busireddy  you might have a bit better understanding of . We noticed `close()` methods below.  Do we need to call in the stop block  ?

```registries.forEach(MeterRegistry::close);
    METER_REGISTRY.close();
```

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java (hunk 2 above)
Comment (buchi-busireddy) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:322]: @samarth-gupta-traceable can't you use iterator and do `itr.remove()` instead?

File: platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java (hunk 3 above)
Comment (buchi-busireddy) [reply platform-metrics/src/main/java/org/hypertrace/core/serviceframework/metrics/PlatformMetricsRegistry.java:319]: So, if we're not really stopping and re-initializing the `PlatformMetricRegistry` in a JVM, the right thing to do is just call `METER_REGISTRY.close()` and that will take care of closing all registered meter registries. 
However, I think we reuse the `METER_REGISTRY` static instance (in unit tests), hence this method was implemented accordingly. Actually, we should feel free to fix that as well.

**Pull request merged**